In [1]:
import pandas as pd
import numpy as np
data_frame = pd.read_csv("Glassdoor_jobs.csv")

In [2]:
data_frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5578 entries, 0 to 5577
Data columns (total 12 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Job Title          5578 non-null   object 
 1   Salary Estimate    5578 non-null   object 
 2   Job Description    5578 non-null   object 
 3   Rating             5578 non-null   float64
 4   Company Name       5578 non-null   object 
 5   Location           5578 non-null   object 
 6   Size               5578 non-null   object 
 7   Founded            5578 non-null   int64  
 8   Type of ownership  5578 non-null   object 
 9   Industry           5578 non-null   object 
 10  Sector             5578 non-null   object 
 11  Revenue            5578 non-null   object 
dtypes: float64(1), int64(1), object(10)
memory usage: 523.1+ KB


In [3]:
data_frame.head(50)

Job Title  \
0             Geospatial Researcher/Data Scientist IV   
1                               Senior Data Scientist   
2   Software Integrator / Data Scientist - Expert ...   
3                                 Lead Data Scientist   
4                                      Data Scientist   
5                                      Data Scientist   
6                           Defense Data Scientist II   
7                                 Senior Data Analyst   
8                                      Data Scientist   
9                     Defense Senior Data Scientist I   
10                                     Data Scientist   
11                                     Data Scientist   
12                 2023771 Data Scientist $215,000.00   
13                              DATA SCIENTIST - USMS   
14                                Junior Data Labeler   
15                                     Data Scientist   
16                     Tech Excellence Data Scientist   
17                                     Data Scientist   
18                              Senior Data Scientist   
19                                     Data Scientist   
20               Jr. Data Scientist (Remote Eligible)   
21                                   Data Scientist I   
22                                     Data Scientist   
23                              Senior Data Scientist   
24               Data Scientist - Electricity Markets   
25                Data Scientist - Clearance Required   
26                      Data Scientist with Polygraph   
27                                     Data Scientist   
28             Senior Data Scientist (Remote/On Site)   
29                                     Data Scientist   
30                         Defense Data Scientist III   
31                         Defense Data Scientist III   
32  Pharmaceutical/Biotech Market Research Data Sc...   
33                             Data Scientist, Junior   
34                                     Data Scientist   
35   Jr. Machine Learning Data Scientist - Python / R   
36                                     Data Scientist   
37              Tech Excellence Data Engineer, Junior   
38                                     Data Scientist   
39                             Data Scientist/Analyst   
40                        Data Scientist (ALL LEVELS)   
41                                     Data Scientist   
42                                     Data Scientist   
43                              Junior Data Scientist   
44                    Data Scientist / Data Analytics   
45                                     Data Scientist   
46                            Aviation Data Scientist   
47                                     Data Scientist   
48      Data Scientist (Hybrid) (Intelligence Center)   
49                                     Data Scientist   

                                Salary Estimate  \
0                 $76K - $120K (Glassdoor est.)   
1                $101K - $138K (Glassdoor est.)   
2                 $74K - $109K (Glassdoor est.)   
3                $116K - $161K (Glassdoor est.)   
4                 $76K - $108K (Glassdoor est.)   
5         Employer Provided Salary:$65K - $205K   
6                 $87K - $126K (Glassdoor est.)   
7        Employer Provided Salary:$100K - $125K   
8        Employer Provided Salary:$125K - $185K   
9                 $96K - $137K (Glassdoor est.)   
10       Employer Provided Salary:$160K - $200K   
11       Employer Provided Salary:$110K - $115K   
12               Employer Provided Salary:$215K   
13        Employer Provided Salary:$50K - $300K   
14                 $46K - $73K (Glassdoor est.)   
15       Employer Provided Salary:$120K - $130K   
16                $83K - $125K (Glassdoor est.)   
17        Employer Provided Salary:$62K - $117K   
18        Employer Provided Salary:$96K - $237K   
19                $91K - $131K (Glassdoor est.)   
20                 $48K - $74K (Glassdoor est.)   
21            

In [4]:
#drop duplicate job postings 
data_frame.drop_duplicates(inplace = True)
data_frame.shape

(3440, 12)

Number of rows got reduced from 5578 to 3440 after removing the duplictaes

In [5]:
# to check if there are any null values in Salary column
data_frame['Salary Estimate'].isna().sum()

0

In [6]:
# checking for types of values Salary Estimate column holds
data_frame['Salary Estimate'].value_counts()

$135K - $175K (Employer est.)             155
-1                                        134
$65K - $205K (Employer est.)              110
$180K - $260K (Employer est.)              86
$210K - $304K (Employer est.)              76
                                         ... 
Employer Provided Salary:$140K - $145K      1
$77K - $108K (Glassdoor est.)               1
$67K - $97K (Glassdoor est.)                1
$78K - $113K (Glassdoor est.)               1
$112K - $157K (Glassdoor est.)              1
Name: Salary Estimate, Length: 810, dtype: int64

In [7]:
# deleting 134 rows with no salary information (-1 values). This is around 0.04% of the total number of rows
data_frame = data_frame[data_frame['Salary Estimate'] != '-1']
data_frame.shape

(3306, 12)

In [8]:
data_frame['Salary Estimate'].value_counts()

$135K - $175K (Employer est.)             155
$65K - $205K (Employer est.)              110
$180K - $260K (Employer est.)              86
$210K - $304K (Employer est.)              76
$120K - $140K (Employer est.)              67
                                         ... 
Employer Provided Salary:$140K - $145K      1
$77K - $108K (Glassdoor est.)               1
$67K - $97K (Glassdoor est.)                1
$78K - $113K (Glassdoor est.)               1
$112K - $157K (Glassdoor est.)              1
Name: Salary Estimate, Length: 809, dtype: int64

In [9]:
# Creating separate columns for following types of salary 
# 1. Per Hour
# 2. Employer Provided Salary
# 3. Employer est.
data_frame['hourly'] = data_frame['Salary Estimate'].apply(lambda x: 1 if 'per hour' in x.lower() else 0)
data_frame['employer provided'] = data_frame['Salary Estimate'].apply(lambda x: 1 if 'employer provided salary' in x.lower() else 0)
data_frame['employer estimate'] = data_frame['Salary Estimate'].apply(lambda x: 1 if 'employer est.' in x.lower() else 0)

In [10]:
# to check if there are values in the new columns
data_frame.loc[data_frame['hourly']==1]

Job Title  \
23                                Senior Data Scientist   
121                                      Data Scientist   
216                     Data Scientist/Big Data Analyst   
226                                      Data Scientist   
306                           Sr. Research Data Analyst   
1020                                     Data Scientist   
1112                  Analytics Developer, Data Science   
1128                     Athletics Gameday Statistician   
1203   Artificial Intelligence Documentation Specialist   
1204      Data Scientist(Data Bricks experience)-Hybrid   
1303                                  Data Scientist IV   
1310                                     Data Scientist   
1311                                     Data Scientist   
1314                                     Data Scientist   
1780                                Senior Data Analyst   
1874                Data Scientist /Data Analyst--H4551   
1888                                 Game Mathematician   
1971           Data Analyst - Senior - Decision Support   
2188  Senior Health Care Research / Senior Healthcar...   
2340                Basketball Scorekeeper/Statistician   
2343                                     Data Scientist   
2501                             Data Analyst/Scientist   
2503                                  Data Scientist II   
2582  Federal Work Study Student - Athletics Statist...   
2789                                       Statistician   
2791                           IT Data Scientist (NC48)   
2794                   Chemical Exposure Data Scientist   
2795                                     Data Scientist   
2802                   Data Scientist, Machine Learning   
2803                                     DATA SCIENTIST   
2805                   Chemical Exposure Data Scientist   
2811                   Data Scientist, Machine Learning   
2820                          Machine Learning Engineer   
2881                                     Data Scientist   
3089                                       Data Analyst   
3090                           Associate Data Scientist   
3288                                     Data Scientist   
3584                                     Data Scientist   
4581                         Digital Data Scientist (H)   
4599                            Data Scientist [Remote]   

                                   Salary Estimate  \
23     Employer Provided Salary:$70 - $75 Per Hour   
121    Employer Provided Salary:$75 - $80 Per Hour   
216    Employer Provided Salary:$50 - $60 Per Hour   
226    Employer Provided Salary:$65 - $70 Per Hour   
306    Employer Provided Salary:$28 - $38 Per Hour   
1020   Employer Provided Salary:$45 - $65 Per Hour   
1112   Employer Provided Salary:$35 - $52 Per Hour   
1128         Employer Provided Salary:$25 Per Hour   
1203   Employer Provided Salary:$20 - $23 Per Hour   
1204   Employer Provided Salary:$35 - $46 Per Hour   
1303         Employer Provided Salary:$84 Per Hour   
1310   Employer Provided Salary:$70 - $75 Per Hour   
1311         Employer Provided Salary:$85 Per Hour   
1314   Employer Provided Salary:$70 - $80 Per Hour   
1780         Employer Provided Salary:$57 Per Hour   
1874   Employer Provided Salary:$40 - $45 Per Hour   
1888   Employer Provided Salary:$20 - $59 Per Hour   
1971   Employer Provided Salary:$43 - $67 Per Hour   
2188   Employer Provided Salary:$33 - $50 Per Hour   
2340   Employer Provided Salary:$13 - $15 Per Hour   
2343   Employer Provided Salary:$65 - $70 Per Hour   
2501         Employer Provided Salary:$40 Per Hour   
2503   Employer Provided Salary:$57 - $65 Per Hour   
2582         Employer Provided Salary:$13 Per Hour   
2789   Employer Provided Salary:$33 - $39 Per Hour   
2791   Employer Provided Salary:$58 - $63 Per Hour   
2794         Employer Provided Salary:$31 Per Hour   
2795   Employer Provided Salary:$70 - $75 Per Hour   
2802         Employer Provided Salary:$50 Per Hour   
2803  

We observed all the rows with string 'per hour' also contain string 'employer provided salary'.

In [11]:
# Removing text after bracket from salary column
salary = data_frame['Salary Estimate'].apply(lambda x: x.split('(')[0])

In [12]:
salary.head(40)

0                                   $76K - $120K 
1                                  $101K - $138K 
2                                   $74K - $109K 
3                                  $116K - $161K 
4                                   $76K - $108K 
5           Employer Provided Salary:$65K - $205K
6                                   $87K - $126K 
7          Employer Provided Salary:$100K - $125K
8          Employer Provided Salary:$125K - $185K
9                                   $96K - $137K 
10         Employer Provided Salary:$160K - $200K
11         Employer Provided Salary:$110K - $115K
12                 Employer Provided Salary:$215K
13          Employer Provided Salary:$50K - $300K
14                                   $46K - $73K 
15         Employer Provided Salary:$120K - $130K
16                                  $83K - $125K 
17          Employer Provided Salary:$62K - $117K
18          Employer Provided Salary:$96K - $237K
19                                  $91K - $131K 


In [13]:
# Removing K and $ from the salary column
minus_Kd = salary.apply(lambda x: x.replace('K','').replace('$',''))
minus_Kd.head(50)

0                                     76 - 120 
1                                    101 - 138 
2                                     74 - 109 
3                                    116 - 161 
4                                     76 - 108 
5             Employer Provided Salary:65 - 205
6                                     87 - 126 
7            Employer Provided Salary:100 - 125
8            Employer Provided Salary:125 - 185
9                                     96 - 137 
10           Employer Provided Salary:160 - 200
11           Employer Provided Salary:110 - 115
12                 Employer Provided Salary:215
13            Employer Provided Salary:50 - 300
14                                     46 - 73 
15           Employer Provided Salary:120 - 130
16                                    83 - 125 
17            Employer Provided Salary:62 - 117
18            Employer Provided Salary:96 - 237
19                                    91 - 131 
20                                     4

In [14]:
# to check how many rows still contains any alphabet values 
minus_Kd.str.contains('[a-zA-Z]').sum()

291

In [15]:
# Number of rows with word 'Employer'
minus_Kd.str.contains('Employer').sum()

291

All the rows with text have word 'Employer' in them. Removing text from the employer rows would remove text from all the rows.

In [16]:
# replacing 'per hour' and 'employer provided salary' with blank 
min_hr = minus_Kd.apply(lambda x: x.lower().replace('per hour','').replace('employer provided salary:',''))

In [17]:
# Creating min_and max salary columns
data_frame['min_salary'] = min_hr.apply(lambda x: int(x.split('-')[0]) if '-' in x else np.NaN)
data_frame['max_salary'] = min_hr.apply(lambda x: int(x.split('-')[1]) if '-' in x else np.NaN)

In [18]:
#Converting hourly wage to annual 
# Multiplying with 2 as 8hrs/day * 5 days/week * 50 weeks=2000
# The year has 52 weeks but 2 of those weeks are assigned to holidays and vacation time.

# As all the salaries are in thousands(000), multiplying only with 2 and not 2000.

data_frame['min_salary'] = data_frame.apply(lambda x: x.min_salary*2 if x.hourly ==1 else x.min_salary, axis =1)
data_frame['max_salary'] = data_frame.apply(lambda x: x.max_salary*2 if x.hourly ==1 else x.max_salary, axis =1)


In [19]:
# to check how do the below columns look like after the conversion
data_frame[data_frame.hourly ==1][['hourly','min_salary','max_salary']]

hourly  min_salary  max_salary
23         1       140.0       150.0
121        1       150.0       160.0
216        1       100.0       120.0
226        1       130.0       140.0
306        1        56.0        76.0
1020       1        90.0       130.0
1112       1        70.0       104.0
1128       1         NaN         NaN
1203       1        40.0        46.0
1204       1        70.0        92.0
1303       1         NaN         NaN
1310       1       140.0       150.0
1311       1         NaN         NaN
1314       1       140.0       160.0
1780       1         NaN         NaN
1874       1        80.0        90.0
1888       1        40.0       118.0
1971       1        86.0       134.0
2188       1        66.0       100.0
2340       1        26.0        30.0
2343       1       130.0       140.0
2501       1         NaN         NaN
2503       1       114.0       130.0
2582       1         NaN         NaN
2789       1        66.0        78.0
2791       1       116.0       126.0
2794       1         NaN         NaN
2795       1       140.0       150.0
2802       1         NaN         NaN
2803       1       126.0       130.0
2805       1         NaN         NaN
2811       1         NaN         NaN
2820       1       140.0       160.0
2881       1        72.0       126.0
3089       1        80.0        90.0
3090       1        60.0        68.0
3288       1        78.0        80.0
3584       1       100.0       120.0
4581       1       100.0       110.0
4599       1       180.0       220.0

In [20]:
# Calculating average salary
data_frame['avg_salary'] =(data_frame.min_salary+data_frame.max_salary)/2

In [21]:
#If salary is not provided in range, assigned the single value salary available
data_frame['avg_salary'][data_frame['avg_salary'].isna()] = min_hr

C:\Users\WM796GU\AppData\Local\Temp\ipykernel_22432\3051027290.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_frame['avg_salary'][data_frame['avg_salary'].isna()] = min_hr


In [22]:
#Converting hourly wage to annual when salary is not provided in range. Considering this only provided value to be average
data_frame['avg_salary'][data_frame['min_salary'].isna() & data_frame.hourly ==1] = data_frame['avg_salary'].astype(float)*2

C:\Users\WM796GU\AppData\Local\Temp\ipykernel_22432\885808434.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_frame['avg_salary'][data_frame['min_salary'].isna() & data_frame.hourly ==1] = data_frame['avg_salary'].astype(float)*2


In [23]:
data_frame[data_frame.hourly ==1][['hourly','min_salary','max_salary','avg_salary']]

hourly  min_salary  max_salary avg_salary
23         1       140.0       150.0      145.0
121        1       150.0       160.0      155.0
216        1       100.0       120.0      110.0
226        1       130.0       140.0      135.0
306        1        56.0        76.0       66.0
1020       1        90.0       130.0      110.0
1112       1        70.0       104.0       87.0
1128       1         NaN         NaN       50.0
1203       1        40.0        46.0       43.0
1204       1        70.0        92.0       81.0
1303       1         NaN         NaN      168.0
1310       1       140.0       150.0      145.0
1311       1         NaN         NaN      170.0
1314       1       140.0       160.0      150.0
1780       1         NaN         NaN      114.0
1874       1        80.0        90.0       85.0
1888       1        40.0       118.0       79.0
1971       1        86.0       134.0      110.0
2188       1        66.0       100.0       83.0
2340       1        26.0        30.0       28.0
2343       1       130.0       140.0      135.0
2501       1         NaN         NaN       80.0
2503       1       114.0       130.0      122.0
2582       1         NaN         NaN       26.0
2789       1        66.0        78.0       72.0
2791       1       116.0       126.0      121.0
2794       1         NaN         NaN       62.0
2795       1       140.0       150.0      145.0
2802       1         NaN         NaN      100.0
2803       1       126.0       130.0      128.0
2805       1         NaN         NaN       62.0
2811       1         NaN         NaN      126.0
2820       1       140.0       160.0      150.0
2881       1        72.0       126.0       99.0
3089       1        80.0        90.0       85.0
3090       1        60.0        68.0       64.0
3288       1        78.0        80.0       79.0
3584       1       100.0       120.0      110.0
4581       1       100.0       110.0      105.0
4599       1       180.0       220.0      200.0

In [24]:
#state of the job location
data_frame['job_state'] = data_frame['Location'].apply(lambda x: x.split(',')[1])

IndexError: list index out of range

In [25]:
# Figuring out reason for the error
# Filtererd the rows which don't have ','
data_frame['Location'][~data_frame['Location'].str.contains(',')].value_counts()

United States    17
Manhattan        17
California       15
Texas             3
St Louis Park     2
Florida           2
Hawaii            1
Pennsylvania      1
Name: Location, dtype: int64

Rows which have above locations do not contain character ',' in them. This is leading to the above error in accessing the second item of the list.

In [26]:
# storing locations which don't have ',' in a 'states' variable so that they can be treated separately
states = data_frame['Location'][~data_frame['Location'].str.contains(',')].value_counts().index

In [27]:
states

Index(['United States', 'Manhattan', 'California', 'Texas', 'St Louis Park',
       'Florida', 'Hawaii', 'Pennsylvania'],
      dtype='object')

In [28]:
# Extracted states from the locations without error this time
data_frame['job_state'] = data_frame['Location'].apply(lambda x: x if x in states else x.split(',')[1])

In [29]:
data_frame['job_state'].value_counts()

 CA              610
 TX              278
 OH              253
 FL              253
 NC              182
 NY              163
 MO              106
 MD              106
 VA              102
 MA               97
 AZ               90
 MI               86
 CO               83
 IL               82
 WA               82
 UT               80
 OR               80
 GA               76
 MN               76
 PA               63
 IN               59
 TN               54
 DC               47
 QC               36
 NV               34
 KS               24
 NJ               18
 HI               17
United States     17
Manhattan         17
California        15
 DE                6
 Cuyahoga          3
Texas              3
St Louis Park      2
Florida            2
Hawaii             1
Pennsylvania       1
 KY                1
 SC                1
Name: job_state, dtype: int64

In [30]:
# We noticed there are leading zeroes in the above output. Removing those leading zeros
data_frame['job_state'] = data_frame['job_state'].apply(lambda x: x.lstrip(' '))

In [31]:
# We noticed there is a state 'QC' in the above output. Removing QC(Quebec), Canada jobs
data_frame = data_frame[data_frame['job_state']!='QC']

In [32]:
data_frame['job_state'].value_counts()

CA               610
TX               278
FL               253
OH               253
NC               182
NY               163
MD               106
MO               106
VA               102
MA                97
AZ                90
MI                86
CO                83
WA                82
IL                82
OR                80
UT                80
GA                76
MN                76
PA                63
IN                59
TN                54
DC                47
NV                34
KS                24
NJ                18
HI                17
United States     17
Manhattan         17
California        15
DE                 6
Cuyahoga           3
Texas              3
St Louis Park      2
Florida            2
Hawaii             1
Pennsylvania       1
KY                 1
SC                 1
Name: job_state, dtype: int64

In [33]:
# Mapping states which didn't have ',' with 2 letter abbreviation 
data_frame['job_state'][data_frame['job_state']=='Manhattan']='NY'
data_frame['job_state'][data_frame['job_state']=='California']='CA'
data_frame['job_state'][data_frame['job_state']=='Texas']='TX'
data_frame['job_state'][data_frame['job_state']=='St Louis Park']='MN'
data_frame['job_state'][data_frame['job_state']=='Florida']='FL'
data_frame['job_state'][data_frame['job_state']=='Hawaii']='HI'
data_frame['job_state'][data_frame['job_state']=='Pennsylvania']='PA'
data_frame['job_state'][data_frame['job_state']=='Cuyahoga']='OH'

C:\Users\WM796GU\AppData\Local\Temp\ipykernel_22432\3094817763.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_frame['job_state'][data_frame['job_state']=='California']='CA'
C:\Users\WM796GU\AppData\Local\Temp\ipykernel_22432\3094817763.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_frame['job_state'][data_frame['job_state']=='Texas']='TX'
C:\Users\WM796GU\AppData\Local\Temp\ipykernel_22432\3094817763.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.h

In [34]:
data_frame['job_state'].value_counts()

CA               625
TX               281
OH               256
FL               255
NC               182
NY               180
MD               106
MO               106
VA               102
MA                97
AZ                90
MI                86
CO                83
IL                82
WA                82
UT                80
OR                80
MN                78
GA                76
PA                64
IN                59
TN                54
DC                47
NV                34
KS                24
NJ                18
HI                18
United States     17
DE                 6
KY                 1
SC                 1
Name: job_state, dtype: int64

In [35]:
data_frame['job_state'].isnull().sum()

0

In [36]:
## Looking for types of job titles
data_frame['Job Title'].value_counts()

Data Scientist                                        498
Senior Data Scientist                                  97
Senior Data Analyst                                    60
Machine Learning Engineer                              43
Senior Associate, Data Analyst Learning Deployment     28
                                                     ... 
Senior Data Scientist, Computer Graphics                1
Sr. eCommerce Data Analyst                              1
Data Analyst – Digital Advertisement                    1
Lead Data Scientist, Digital Product Development        1
Machine Learning Principal/Architect                    1
Name: Job Title, Length: 1333, dtype: int64

Although, my scrapper code searched only for 'data scientist' roles on Glassdoor, data got pulled for many other types of roles. Let's divide them in broad categories.

Also, there are many types of data scientist roles. Ex: Computer Vision Data Scientist. I am categorizing all those data scientist roles into 1 data scientist category.

In [37]:
def title_simplifier(title):
    if 'data scientist' in title.lower():
        return 'data scientist'
    elif 'data engineer' in title.lower():
        return 'data engineer'
    elif 'analyst' in title.lower():
        return 'analyst'
    elif 'machine learning' in title.lower():
        return 'mle'
    elif 'manager' in title.lower():
        return 'manager'
    elif 'director' in title.lower():
        return 'director'
    else:
        return 'na'

In [38]:
data_frame['job_simp'] = data_frame['Job Title'].apply(title_simplifier)
data_frame['job_simp'].value_counts()

data scientist    1666
na                 574
analyst            504
mle                254
data engineer      120
manager            101
director            51
Name: job_simp, dtype: int64

For the 'NA' values, even if they are not true data scientist positions, they can still be used for the salary prediciton.

In [39]:
# defining seniority
def seniority(title):
    if 'sr' in title.lower() or 'senior' in title.lower() or 'lead' in title.lower() or 'principal' in title.lower():
            return 'senior'
    elif 'jr' in title.lower() or 'junior' in title.lower() or 'entry level' in title.lower():
        return 'junior'
    else:
        return 'na'

In [40]:
data_frame['seniority'] = data_frame['Job Title'].apply(seniority)
data_frame.seniority.value_counts()

na        2077
senior    1135
junior      58
Name: seniority, dtype: int64

In [41]:
#  Job description length 
data_frame['desc_len'] = data_frame['Job Description'].apply(lambda x: len(x))
data_frame['desc_len']

0        984
1       1127
2        594
3        853
4       1325
        ... 
5343     709
5344     817
5351     884
5374     581
5427    2125
Name: desc_len, Length: 3270, dtype: int64

In [42]:
#age of company 
data_frame['age'] = data_frame.Founded.apply(lambda x: x if x <1 else 2022 - x)

In [43]:
data_frame.head()

Job Title  \
0            Geospatial Researcher/Data Scientist IV   
1                              Senior Data Scientist   
2  Software Integrator / Data Scientist - Expert ...   
3                                Lead Data Scientist   
4                                     Data Scientist   

                  Salary Estimate  \
0   $76K - $120K (Glassdoor est.)   
1  $101K - $138K (Glassdoor est.)   
2   $74K - $109K (Glassdoor est.)   
3  $116K - $161K (Glassdoor est.)   
4   $76K - $108K (Glassdoor est.)   

                                     Job Description  Rating  \
0  Battelle delivers when others can’t. We conduc...     3.6   
1  JDSAT, Inc. is a solutions company - utilizing...     4.3   
2  *** Must have TS/SCI ***\nJOB DESCRIPTION:\nIn...     4.6   
3  Tensure is a fully-remote technology consultin...     3.8   
4  JDSAT, Inc. is a solutions company - utilizing...     4.3   

           Company Name          Location                    Size  Founded  \
0              Battelle  Fort Belvoir, VA  1001 to 5000 Employees     1929   
1                 JDSAT        McLean, VA       1 to 50 Employees       -1   
2  Que Technology Group        McLean, VA       1 to 50 Employees       -1   
3    Tensure Consulting    Washington, DC       1 to 50 Employees       -1   
4                 JDSAT        McLean, VA       1 to 50 Employees       -1   

        Type of ownership                Industry  ... employer provided  \
0  Nonprofit Organization  Research & Development  ...                 0   
1       Company - Private                      -1  ...                 0   
2       Company - Private                      -1  ...                 0   
3       Company - Private                      -1  ...                 0   
4       Company - Private                      -1  ...                 0   

  employer estimate  min_salary  max_salary  avg_salary  job_state  \
0                 0        76.0       120.0        98.0         VA   
1                 0       101.0       138.0       119.5         VA   
2                 0        74.0       109.0        91.5         VA   
3                 0       116.0       161.0       138.5         DC   
4                 0        76.0       108.0        92.0         VA   

         job_simp seniority desc_len age  
0  data scientist        na      984  93  
1  data scientist    senior     1127  -1  
2  data scientist        na      594  -1  
3  data scientist    senior      853  -1  
4  data scientist        na     1325  -1  

[5 rows x 23 columns]

In [44]:
##parsing of job description (python, etc.)
#python
data_frame['python_yn'] = data_frame['Job Description'].apply(lambda x: 1 if 'python' in x.lower() else 0)
data_frame.python_yn.value_counts()

0    2981
1     289
Name: python_yn, dtype: int64

In [45]:
#SQL
data_frame['sql_yn'] = data_frame['Job Description'].apply(lambda x: 1 if 'sql' in x.lower() else 0)
data_frame.sql_yn.value_counts()

0    3078
1     192
Name: sql_yn, dtype: int64

In [46]:
data_frame.to_csv('salary_data_cleaned.csv', index = False)